### ISSUE TITLE
firedrake-install failure on Debian stretch

### ISSUE THREAD

**Issue by gdmcbain** [firedrakeproject/firedrake#904]
Firedrake has been running for a while on this system, 64-bit Debian GNU/Linux testing (stretch), with regular apt upgrades and firedrake-updates.   A week or so ago the latter failed, presumably following something in the former (libopenmpi2?), so an attempt was made to freshly reinstall.  This failed.  Following the [instructions](http://www.firedrakeproject.org/download.html), I reran with `--log` and am attaching the [result](https://github.com/firedrakeproject/firedrake/files/512516/firedrake-install.log.txt). I had a look myself but wasn't able to make too much of it.

**Comment by stephankramer** [comment]
Looks like it's failing to build mumps (which is built as part of petsc). According to http://lists.mcs.anl.gov/pipermail/petsc-dev/2015-September/018229.html, this may happen if during its build it is picking up a include file from an older mumps build. Do you have the libmumps-dev package installed by any chance? This seems to install /usr/include/cmumps_struc.h which may conflict with your build.

**Comment by gdmcbain** [comment]
I do, "Version : 4.10.0.dfsg-4+b1".

It's required by (as shown by `sudo aptitude purge libmumps-dev`): 

> 1)      dolfin-bin [2016.1.0-5 (now, testing)]
> 2)      dolfin-doc [2016.1.0-5 (now, testing)]
> 3)      fenics [1:1.5.0.1 (now, testing)]
> 4)      libdolfin-dev [2016.1.0-5 (now, testing)]
> 5)      libgmm++-dev [5.1+dfsg1-3 (now, testing)]
> 6)      libpetsc3.7-dev [3.7.3.dfsg1-3 (now, testing)]
> 7)      libpetsc3.7.3-dev [3.7.3.dfsg1-3 (now, testing)]
> 8)      libslepc3.7-dev [3.7.2+dfsg1-3 (now, testing)]
> 9)      libslepc3.7.2-dev [3.7.2+dfsg1-3 (now, testing)]
> 10)     python-dolfin [2016.1.0-5 (now, testing)]

**Comment by gdmcbain** [comment]
I wasn't all that keen on purging libmumps-dev (e.g. I wanted PETSc for Gmsh to generate the meshes to pass to Firedrake), but, figuring that I should be able to reinstall it later, I did.

With that, `firedrake-install` seemed to work, as shown in the attached 
[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/515114/firedrake-install.log.txt), but I found that my old scripts didn't work (ImportErrors for mpi4py or petsc4py).  I won't go into that because, continuing to follow the [instructions](http://www.firedrakeproject.org/download.html), I ran the test suite, which failed similarly

> geordie@laplace:~/src/firedrake/src/firedrake$ . ../../bin/activate
> (firedrake)geordie@laplace:~/src/firedrake/src/firedrake$ make alltest
>    Building extension modules
>    Linting firedrake codebase
>    Linting firedrake test suite
>    Linting firedrake scripts
>    Running all regression tests
>Traceback (most recent call last):
>  File "/usr/local/users/geordie/src/firedrake/local/lib/python2.7/site-packages/_pytest/config.py", line 320, in _importconftest
>    mod = conftestpath.pyimport()
>  File "/usr/local/users/geordie/src/firedrake/local/lib/python2.7/site-packages/py/_path/local.py", line 650, in pyimport
>    __import__(modname)
>  File "/usr/local/users/geordie/src/firedrake/src/firedrake/tests/conftest.py", line 5, in <module>
>    from mpi4py import MPI
> ImportError: libmpi.so.12: cannot open shared object file: No such file or directory
> ERROR: could not load /usr/local/users/geordie/src/firedrake/src/firedrake/tests/conftest.py

Indeed I do not have libmpi.so.12.  Courtesy of APT package libopenmpi2, I have /usr/lib/openmpi/lib/libmpi.so.20.0.1 .

**Comment by gdmcbain** [comment]
I note that outside Firedrake's virtualenv, mpi4py works fine.

>(firedrake)geordie@laplace:~/src$ deactivate 
>geordie@laplace:~/src$ mpirun -np 4 python -c 'from mpi4py import MPI; print MPI.COMM_WORLD.Get_rank()'
> 1
> 3
> 0
> 2
> geordie@laplace:~/src$ . firedrake/bin/activate
> (firedrake)geordie@laplace:~/src$ mpirun -np 4 python -c 'from mpi4py import MPI; print MPI.COMM_WORLD.Get_rank()'
> Traceback (most recent call last):
>  File "<string>", line 1, in <module>
> ImportError: libmpi.so.12: cannot open shared object file: No such file or directory
> Traceback (most recent call last):
>  File "<string>", line 1, in <module>
>ImportError: libmpi.so.12: cannot open shared object file: No such file or directory
>-------------------------------------------------------
>Primary job  terminated normally, but 1 process returned
>a non-zero exit code.. Per user-direction, the job has been aborted.
>-------------------------------------------------------
>--------------------------------------------------------------------------
>mpirun detected that one or more processes exited with non-zero status, thus causing
>the job to be terminated. The first process to do so was:

  Process name: [[16802,1],0]
  Exit code:    1
--------------------------------------------------------------------------

**Comment by stephankramer** [comment]
About the mumps thing: I'm not sure there's much that can be done. Firedrake installs everything in its own environment precisely because of possible conflicts with different versions of packages, etc. However in this case that mumps package installs something in /usr/include that can't really be ignored by the firedrake build. My suggestions:

1. Build and use Dolfin within a container via docker. This is what I do. Installing or upgrading it as a package tends to break other things on my machine. I think this is now also the recommended way of installing Dolfin.
2. Built firedrake without mumps support by building it with the --minimal-petsc option

About the mpi4py thing, is this not the same issue you had last time (out of date pip cache) in issue #722. If you've done a serious upgrade, I would get rid of your entire ~/.pip/cache/ directory, as pip has no way of figuring out which binary builds are still valid.

**Comment by dham** [comment]
For mpi4py, an alternative is, with the firedrake virtualenv active:
````
pip uninstall mpi4py
pip install --no-binary mpi4py mpi4py
````

**Comment by gdmcbain** [comment]
Thank you both for the suggestions.

@stephankramer : That's fair enough for a user-level package like Dolfin, but the problem isn't so much Dolfin as PETSc, which is needed by many other packages.  I used to keep PETSc up to date myself, but would prefer to leave that to the operating system's package management if possible.

I'll keep the `--minimal-petsc` option in mind, but wonder how limiting that would be in terms of options for preconditioning, etc.; easy access to all those PETSc options is one of the great strengths of Firedrake.

@dham : Thanks, that fixed mpi4py.  The above one-liner now works inside and outside Firedrake's virtualenv and my old scripts run again.

To summarize: I purged libmumps-dev via `aptitude`, which also removed `libpetsc3.7.3-dev`, breaking my Gmsh and much else, then I reinstalled Firedrake in the usual way and inside Firedrake's virtualenv uninstalled and reinstalled mpi4py with `pip` and `--no-binary` as suggested by David, then I reinstalled libpetsc3.7.3-dev with `aptitude`.  Everything is now playing together.  (Not sure about Dolfin, which I don't really use.  I had only installed it quickly to run an example that someone had posted on the Firedrake mailing list.  If I need Dolfin again, I'll follow up Stephan's suggestion of keeping it in a container.)
